src/padding.jl

```diff
@@ -0,0 +1,205 @@
+on every dimension in `dims`. In this case, `dims` 
+defaults to the first `ndims(x)-2` dimension 
+(i.e. excludes the channel and batch dimension).
+
+See also [`pad_reflect`](@ref) and [`pad_repeat`](@ref).
+
+```jldoctest
+julia> pad_constant(reshape(1:4, 2, 2), (1, 2, 3, 4), 8)
+5×9 Matrix{Int64}:
+ 8  8  8  8  8  8  8  8  8
+ 8  8  8  1  3  8  8  8  8
+ 8  8  8  2  4  8  8  8  8
+ 8  8  8  8  8  8  8  8  8
+ 8  8  8  8  8  8  8  8  8
+````
+"""
+function pad_constant(x::AbstractArray, pad::NTuple{M,Int}, val=0; 
+                    dims=1:M÷2) where M
+  length(dims) == M÷2 ||
+    throw(ArgumentError("The number of dims should be equal to the number of padding dimensions"))
+  outsize, center = compute_pad_outsize_and_center(size(x), pad, dims)
+  y = fill!(similar(x, eltype(x), outsize), val)
+  y[center...] .= x
+  return y
+end
+
+function compute_pad_outsize_and_center(sz::NTuple{N,Int}, pad, dims) where {N}
+  leftpad, rightpad = pad[1:2:end], pad[2:2:end]
+  
+  outsize = ntuple(N) do i
+              k = findfirst(==(i), dims)
+              if k === nothing
+                return sz[i]
+              else
+                return sz[i] + leftpad[k] + rightpad[k]
+              end
+            end::NTuple{N,Int}
+  
+  leftcorner = ones(Int, N)
+  rightcorner = collect(outsize)
+  for (i, d) in enumerate(dims)
+    leftcorner[d] += leftpad[i]
+    rightcorner[d] -= rightpad[i]
+  end
+  center = ntuple(N) do i
+            leftcorner[i]:rightcorner[i]
+          end 
+  return outsize, center
+end
+
+function rrule(::typeof(pad_constant), x::AbstractArray, pad::NTuple{M,Int}, val=0; 
+      dims=1:M÷2) where M
+  szx = size(x)
+  y = pad_constant(x, pad, val; dims=dims)
+  function pad_constant_pullback(Δ)
+    outsize, center = compute_pad_outsize_and_center(szx, pad, dims)
+    (NO_FIELDS, 
+    @thunk(Δ[center...]), 
+    DoesNotExist(),
+    @thunk(sum(Δ) - sum(Δ[center...])), 
+    )
+  end
+  return y, pad_constant_pullback
+end
+
+
+"""
+    pad_repeat(x, pad::Tuple; [dims])
+    pad_repeat(x, pad::Int; [dims])
+ 
+Pad the array `x` repeating the values on the border.
+
+`pad` can a tuple of integers `(l1, r1, ..., ln, rn)`
+of some length `2n` that specifies the left and right padding size
+for each of the dimensions in `dims`. If `dims` is not given, 
+it defaults to the first `n` dimensions.
+
+For integer `pad` input instead, it is applied on both sides
+on every dimension in `dims`. In this case, `dims` 
+defaults to the first `ndims(x)-2` dimensions 
+(i.e. excludes the channel and batch dimension). 
+
+See also [`pad_reflect`](@ref) and [`pad_constant`](@ref).
+
+```jldoctest
+julia> pad_repeat(reshape(1:9, 3, 3), (1,2,3,4))
+6×10 Matrix{Int64}:
+ 1  1  1  1  4  7  7  7  7  7
+ 1  1  1  1  4  7  7  7  7  7
+ 2  2  2  2  5  8  8  8  8  8
+ 3  3  3  3  6  9  9  9  9  9
+ 3  3  3  3  6  9  9  9  9  9
+ 3  3  3  3  6  9  9  9  9  9
+```
+"""
+function pad_repeat(x::AbstractArray, pad::NTuple{M,Int}; 
```

Comment:
I'm inclined not to define this and instead just tell users to use `repeat`. Is there an advantage to defining it here?

Reply:
as you can see by yourself the code is not trivial, you can't tell users "use repeat". So yes, we should have that here. 
Another point is that sometimes we provide some functionality even if the implementation is simple (think about the `Dense` layer) because it is still convenient to have and makes for a unified and consistent experience.

Reply:
Its a different API entirely, and any "non-trivial"-ness in the implementation comes from forcing the same kwarg handling that has come up before. You definitely can tell users, and it would probably be better supported if you did.

With Dense there is the convenience factor, here the API call syntax might be different, but you can get a more complete implementation, which is clearly better since becoming a DSL is clearly the wrong way forward anyway.